### ISSUE TITLE
Programmatically receive progress events

### ISSUE THREAD

**Issue by davej** [electron-userland/electron-builder#3493]
* **Version**:  20.29.0

I would like to consume the progress events of `electron-builder` programmatically. Currently, it looks like all I can do is try to parse the output from `stdout`. Is there a way to consume progress events without resorting to parsing `stdout`? If not, would this be considered as a feature request?

**Comment by develar** [comment]
You can use app-builder-lib module directly and listen artifactCreated event.

https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/remoteBuilder/builder-cli.ts#L52

If you need publish, don't forget to create publish manager — https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/index.ts#L51

**Comment by davej** [comment]
Thank you @develar, really appreciate your pointers. A few more questions which I would love your help on.

1. Is it possible to get something like an `artifactStarted` event?
2. Is it possible to get access to something like `uploadStarted` and `uploadFinished` events from publisher?
3. Is it possible to provide my own logger instance (or somehow override the existing one) so I can parse it? (Instead of trying to parse `stdout` directly).

 If you like, I can submit a PR with some documentation on this when I have a decent understanding?

**Comment by develar** [comment]
@davej To ensure that I don't waste my time, could you please explain what do you want to achieve and for what do you need so tight integration?I want to understand your purpose to provide more concise tips.

**Comment by davej** [comment]
Of course. I'm working on an online tool that converts URLs into fully-built electron desktop apps that are ready for distribution.

The online builder currently reports progress events (and a progress bar) from the builder to the UI using Web Sockets. The current version doesn't use electron-builder and I call different packages (e.g..`electron-winstaller`) directly from my code. As a result of having more control, the progress bar on the UI is pretty granular and accurate.

So, to summarise: I would like to switch to `electron-builder` but still retain the accurate progress events and progress bar if possible.

**Comment by develar** [comment]
I hope you are aware about https://github.com/electron-userland/electron-build-service.

Ok, so, indeed you should use the same approach as electron build service — see first link, https://github.com/electron-userland/electron-builder/blob/master/packages/app-builder-lib/src/remoteBuilder/builder-cli.ts#L74

Do you see? Private method `_build` is used.

`artifactStarted` event...

There are 2 approaches:

A. build exactly as CLI version does.
B. build in pre-packaged format and then build in distributable format.

electron build service uses `B` because of privacy — to build project into pre-packaged format we need all project sources. And then this pre-packaged transferred to remote server.

App can be build in several distributable formats in parallel.

So, what I want to say — if you will use fine-grained control over build, you don't need `artifactStarted` event. At first, you will package it in prepackaged format, then invoke (in parallel or whatever you want) electron-builder again with `prepackaged` option to build in distributable format.

Ok, ok, maybe it is too complicated and if your app does the whole job in one session on one machine, you can simply use app-builder-lib (again, I suggest to not use electron-builder directly programmatically to avoid yargs dependency).

I will add events till tomorrow morning CET.

**Comment by davej** [comment]
> I hope you are aware about https://github.com/electron-userland/electron-build-service.

No, I'm not aware of `electron-build-service` yet. Thanks for the link, I will take a look. Looks very interesting.

> There are 2 approaches:
>
> A. build exactly as CLI version does.
> B. build in pre-packaged format and then build in distributable format.

Ok, I understand. Privacy is not such an issue for me, because I am building from publicly accessible resources anyway (i.e.. the website has to be hosted on a publicly accessible URL). I think eventually I will split out the build process so it can be parallelized more easily. For now, though, I want to keep it simple because I'm on a schedule and I want to rely on well-tested libraries because I have to ensure it works across 3 different build servers: Windows (because of EV hardware dongle signing), Mac and Linux.

> I will add events till tomorrow morning CET.

Thank you, this is much appreciated! Will you be adding them as config hooks or events on packager out of interest?

**Comment by develar** [comment]
artifactBuildStarted and artifactBuildCompleted added (released 3 days ago, forgot to comment here).

**Comment by davej** [comment]
Thanks for this @develar.

Are there plans to add `uploadStarted` and `uploadFinished` events? I'm just asking because otherwise I will build my own custom uploader but it would be nicer to use electron-builder.
